### PR TITLE
WDPD-137 Improve response handling for non-3DS CC transactions in Azure

### PR DESCRIPTION
**Pull request by ps-mstifter**
### This PR

* improves response handling for non-3DS credit card transactions when the Azure theme is activated

### Jira Links

* https://jira.parkside.at/browse/WDPD-137

### CODE REVIEW

File: Core/ResponseHandler.php [hunk 1]
```diff
@@ -55,7 +58,19 @@ public static function onSuccessResponse($oResponse, $oBackendService, $oOrder)
 
         if (!self::_isPostProcessingAction($oResponse)) {
             self::_updateOrder($oOrder, $oResponse, $oBackendService);
-            $oOrder->sendOrderByEmail();
+
+            try {
+                $oOrder->sendOrderByEmail();
+            } catch (SystemComponentException $exc) {
+                // this error occurrs when the 'Azure' theme is activated and a non-3DS credit card transaction is made
```
Comment (nickstamat) [inline Core/ResponseHandler.php]: minor typo: `occurs`

File: Core/ResponseHandler.php [hunk 2]
```diff
@@ -55,7 +58,19 @@ public static function onSuccessResponse($oResponse, $oBackendService, $oOrder)
...
+                // everything was actually successful, but the 'getThumbnailUrl' method does not exist on the BasketItem
+                // in the 'order_cust.tpl' file. There is no need to do anything in this case but normally continue with
+                // the response handling.
+                $oLogger->debug($exc->getMessage());
+            } catch (Exception $exc) {
+                // if it is a different exception, we have to re-throw it
```
Comment (ps-holler) [inline Core/ResponseHandler.php]: Why does this need to be re-thrown? Any uncaught exception will be thrown anyway, right?

File: Core/ResponseHandler.php [hunk 3]
```diff
@@ -55,7 +58,19 @@ public static function onSuccessResponse($oResponse, $oBackendService, $oOrder)
 
         if (!self::_isPostProcessingAction($oResponse)) {
             self::_updateOrder($oOrder, $oResponse, $oBackendService);
-            $oOrder->sendOrderByEmail();
+
+            try {
+                $oOrder->sendOrderByEmail();
+            } catch (SystemComponentException $exc) {
```
Comment (ps-holler) [inline Core/ResponseHandler.php]: Should be `$oException` to stick with Hungarian notation. Same below.

Comment (ps-cgrach) [inline Core/ResponseHandler.php]: we should check that the message contains `getThumbnailUrl `, so we don't just catch all `SystemComponentException`

Comment (ps-mstifter) [reply Core/ResponseHandler.php]: 👍

File: Core/ResponseHandler.php (hunk 1 above)
Comment (ps-mstifter) [reply Core/ResponseHandler.php]: 👍

File: Core/ResponseHandler.php (hunk 3 above)
Comment (ps-mstifter) [reply Core/ResponseHandler.php]: good point 👍

File: Core/ResponseHandler.php (hunk 2 above)
Comment (ps-mstifter) [reply Core/ResponseHandler.php]: 👍

File: Core/ResponseHandler.php [hunk 4]
```diff
@@ -55,7 +56,21 @@ public static function onSuccessResponse($oResponse, $oBackendService, $oOrder)
...
+                $oOrder->sendOrderByEmail();
+            } catch (SystemComponentException $oException) {
+                // this error occurs when the 'Azure' theme is activated and a non-3DS credit card transaction is made
+                // everything was actually successful, but the 'getThumbnailUrl' method does not exist on the BasketItem
+                // in the 'order_cust.tpl' file. There is no need to do anything in this case but normally continue with
+                // the response handling.
+                // Otherwise the exception is re-thrown.
+                $sAzureErrorMsg = "Function 'getThumbnailUrl' does not exist or is not accessible!";
+
+                if (strpos($oException->getMessage(), $sAzureErrorMsg) !== 0) {
```
Comment (nickstamat) [inline Core/ResponseHandler.php]: This is too strict to be robust enough. What if the message is modified and this falsely throws the exception?

Comment (ps-mstifter) [reply Core/ResponseHandler.php]: 👍

**Pull request merged**